oml/functional/metrics.py

```diff
@@ -536,28 +540,31 @@ def calc_pcf(embeddings: FloatTensor, pcf_variance: Tuple[float, ...]) -> List[f
     return metric
 
 
-def extract_pos_neg_dists(
-    distances: FloatTensor, mask_gt: BoolTensor, mask_to_ignore: Optional[BoolTensor] = None
-) -> Tuple[FloatTensor, FloatTensor]:
+def extract_pos_neg_dists(distances: FloatTensor, gt_ids: List[LongTensor]) -> Tuple[FloatTensor, FloatTensor]:
     """
     Extract distances between relevant samples, and distances between non-relevant samples.
 
     Args:
-        distances: Distance matrix with the shape of ``[query_size, gallery_size]``
-        mask_gt: ``(i,j)`` element indicates if for i-th query j-th gallery is the correct prediction
-        mask_to_ignore: Binary matrix to indicate that some elements in gallery cannot be used
-                     as answers and must be ignored
+        distances: Distance matrix with the shape of ``[query_size, top_n_predicted]``
+        gt_ids: Gallery ids relevant to every query, list of ``n_query`` elements where every element may
+            have an arbitrary length. Every element is within the range ``(0, n_gallery - 1)``
+
     Returns:
         pos_dist: Tensor of distances between relevant samples
         neg_dist: Tensor of distances between non-relevant samples
     """
-    if mask_to_ignore is not None:
-        mask_to_not_ignore = ~mask_to_ignore
-        pos_dist = distances[mask_gt & mask_to_not_ignore]
-        neg_dist = distances[~mask_gt & mask_to_not_ignore]
-    else:
-        pos_dist = distances[mask_gt]
-        neg_dist = distances[~mask_gt]
+    assert len(distances) == len(gt_ids), (distances.shape, len(gt_ids))
+
+    n_q, n_retrieved = distances.shape
+    mask_gt = BoolTensor(size=(n_q, n_retrieved))
+
+    # it's not a heavy cycle because number of gts for every query is small
+    for i_query, ids in enumerate(gt_ids):
+        mask_gt[i_query, clip_max(ids, n_retrieved)] = True
```

Comment:
The `ids` variable is of type `LongTensor`, but the first argument of `clip_max` function needs to be a `tuple`. To prevent type checking errors, let's update the type of the first argument of `clip_max`.